### PR TITLE
Split layout

### PR DESCRIPTION
**Pull request by stefanceriu**
This PR implements a split layout for iPad and MacOS
- it introduces a SingleScreenCoordinator which deals with switching from the splash screen to the home screen split view
- also introduces a SplitScreenCoordinator that deals showing other coordinators side by side but also collapsing them into a single navigation stack in compact layouts
- it renames the old NavigationController to StackScreenCoordinator
- it changes the UserSessionFlowCoordinator state machine to allow for rooms to be shown and switched between while the room list is visible. Also allows modals to presented on top of the split view
- fixes a timeline table view reuse problem

https://user-images.githubusercontent.com/637564/206439218-12b4a558-6b66-4d57-bcbb-5db4eeec97e5.mp4

Screenshot 2022-12-08 at 13 50 14

### CODE REVIEW

**Review by pixlwave — COMMENTED** [review]
Some of these comments may be out of date by now, but I've had a look through :)

File: ElementX/Sources/Services/UserSession/UserSessionFlowCoordinator.swift [hunk 1]
```diff
@@ -120,7 +135,9 @@ class UserSessionFlowCoordinator: CoordinatorProtocol {
             }
         }
 
-        navigationController.setRootCoordinator(coordinator)
+        sidebarStackScreenCoordinator.setRootCoordinator(coordinator)
+        
+        rootScreenCoordinator.setRootCoordinator(splitScreenCoordinator)
```
[the remark this answers is not in the document]
Comment (stefanceriu) [reply ElementX/Sources/Services/UserSession/UserSessionFlowCoordinator.swift:140]: So would you move it to `setupStateMachine` -> `case (.initial, .start, .roomList)` ?

File: ElementX/Sources/Application/Navigation/NavigationScreenCoordinators.swift [hunk 2]
```diff
@@ -0,0 +1,572 @@
...
+    /// - Parameter placeholderCoordinator: coordinator to use if no siderbar or detail is set
+    init(placeholderCoordinator: CoordinatorProtocol) {
+        placeholderModule = NavigationModule(placeholderCoordinator)
+    }
+    
+    /// Set the coordinator to be used on the split's left pannel
+    /// - Parameters:
+    ///   - coordinator: the sidebar coordinator
+    ///   - dismissalCallback: called when this particular sidebar coordinator has removed/replaced
+    func setSidebarCoordinator(_ coordinator: (any CoordinatorProtocol)?, dismissalCallback: (() -> Void)? = nil) {
```
[the remark this answers is not in the document]
Comment (stefanceriu) [reply ElementX/Sources/Application/Navigation/NavigationScreenCoordinators.swift:117]: I'm not sure honestly but I figured if the underlying NavigationStack / SplitView allows it we should do to ¯\_(ツ)_/¯

[the remark this answers is not in the document]
Comment (pixlwave) [reply ElementX/Sources/Application/Navigation/NavigationScreenCoordinators.swift]: Fair enough :)

File: ElementX/Sources/Services/UserSession/UserSessionFlowCoordinator.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (pixlwave) [reply ElementX/Sources/Services/UserSession/UserSessionFlowCoordinator.swift]: I had `AppCoordinator.setupUserSession()` in mind. Then it doesn't even require the `rootScreenCoordinator` to even be passed into the `UserSessionFlowCoordinator`.

**Pull request merged**